src/client/signatures/canonical_json/src/ser.rs

```diff
@@ -0,0 +1,486 @@
+    // looking for \u{X} \u{XX}, \u{XXX}, \u{XXXX} to remove the curly braces
+
+    let mut formatted_string = String::new();
+
+    let mut unicode_start_index = 0;
+
+    while unicode_start_index < serialized_string.len() {
+        let mut is_unicode_start = false;
+        // we must try to create a substring of 4 different sizes
+        for size in 6..9 {
+            if unicode_start_index + size > serialized_string.len() { break; }
+            let substring = match serialized_string.get(unicode_start_index..unicode_start_index + size) {
+                Some(string) => string,
+                _ => return "null".to_owned()
+            };
+
+            if &substring[0..3] == r"\u{" && &substring[size - 1..] == "}" {
+                is_unicode_start = true;
+                // add zero padding
+                let mut padded_unicode = String::from(r"\u");
+                padded_unicode.push_str(&std::iter::repeat("0").take(4 - (size - 4)).collect::<String>());
+                padded_unicode.push_str(&substring[3..size-1]);
+                formatted_string.push_str(&padded_unicode);
+                unicode_start_index = unicode_start_index + size - 1;
+            }
+        }
+
+        if !is_unicode_start {
+            let curr_char = match serialized_string.get(unicode_start_index..unicode_start_index + 1) {
+                Some(character) => character,
+                _ => return "null".to_owned()
+            };
+
+            formatted_string.push_str(&curr_char);
+        }
+
+        unicode_start_index += 1;
+    }
+
+    return formatted_string;
+}
+
+/// Serialize a JSON value to String
+/// 
+/// # Examples
+/// ```rust
+/// # use canonical_json::ser::serialize;
+/// # use serde_json::json;
+/// # fn main() {
+///     serialize(&json!(null)); // returns "null"
+/// 
+///     serialize(&json!("test")); // returns "test"
+/// 
+///     serialize(&json!(10.0_f64.powf(21.0))); // returns "1e+21"
+/// 
+///     serialize(&json!({
+///         "a": "a",
+///         "id": "1",
+///         "b": "b"
+///     })); // returns "{"a":"a","b":"b","id":"1"}"; (orders object keys)
+/// 
+///     serialize(&json!(vec!["one", "two", "three"])); // returns "["one","two","three"]"
+/// # }
+/// 
+/// ```
+pub fn serialize(input: &serde_json::Value) -> String {
+
+    let string = vec![];
+    let null_string = "null".to_owned();
+    let mut serializer = serde_json::Serializer::with_formatter(string, JSONFormatter{});
+    match input.serialize(&mut serializer) {
+        Ok(()) => match String::from_utf8(serializer.into_inner()) {
+            Ok(serialized_string) => return serialized_string,
+            Err(error) => {
+                error!("Error converting bytes to string: {}", error);
+                return null_string;
+            }
+        },
+        Err(error) => {
+            error!("Error serializing JSON : {}", error);
+            return null_string;
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Is there any JSON that cannot be represented in the canonical JSON notation?  If so, consider just having a list of strings in canonical JSON form, and then verifying that the roundtrip String -> `json::Value` -> `to_canonical_json_string` preserves the value.  The named functions don't really add much.